articles/20240820_optimizing-gpu-utilization-in-containerized-environments.md

```diff
@@ -0,0 +1,271 @@
+
+```plaintext
+WORKSPACE | ✓ Request submitted
+WORKSPACE | ✓ Creating workspace nanoGPT
+daytona   | Creating project nanoGPT
+daytona   | Pulling image...
+daytona   | Pulling from daytonaio/workspace-project
+...
+daytona   | Project nanoGPT created
+daytona   | Starting project nanoGPT
+daytona   | Project nanoGPT started
+```
+
+#### List Workspaces:
+
+To keep track of all your workspaces:
+
+```bash
+daytona list
+```
+
+This command lists all created workspaces, including their repository, target, and status.
+
+**Tip**: You can open the created nanoGPT workspace in your preferred [IDE](https://www.daytona.io/docs/reference/cli/#daytona-code) `daytona code` command
+
+## Build and Run the Docker Image
+
+### Create a Dockerfile for nanoGPT
+
+Next, create a Dockerfile tailored to run nanoGPT with GPU support. Here’s an example Dockerfile:
+
+```Dockerfile
+FROM nvidia/cuda:11.7.1-cudnn8-devel-ubuntu20.04
+
+# Install Python and necessary dependencies
+
+RUN apt-get update && apt-get install -y python3-pip git
+
+# Copy nanoGPT code into the container
+
+COPY . /nanoGPT
+WORKDIR /nanoGPT
+
+# Install Python dependencies
+
+RUN pip install torch transformers
+
+# Define the entry point for the nanoGPT script
+
+CMD ["python3", "train.py"]
+```
+
+### Build the Docker Image
+
+Build the Docker image for nanoGPT:
+
+```bash
+docker build -t nanogpt-container .
+```
+
+### Run nanoGPT with GPU Support
+
+Run the container, allocating GPU resources for training:
+
+```bash
+docker run --gpus all nanogpt-container
+```
+
+## Common Challenges and Solutions
+
+As you work with GPU-intensive tasks in containerized environments, you may encounter several challenges. Here’s how to address some common issues:
+
+**Problem:** _[The CUDA version inside the container might not match the GPU driver on the host system]_
+
+**Solution:** _[Ensure that the CUDA version in your container is compatible with your host’s NVIDIA driver. Consult NVIDIA’s [compatibility matrix](https://docs.nvidia.com/deploy/cuda-compatibility/index.html) to avoid mismatches]_
+
+**Problem:** _[GPU resources are not being fully utilized, leading to slower performance.]_
+
+**Solution:** _[Profile GPU usage with `nvidia-smi` and optimize your application code. Consider using mixed precision training or inference (fp16) to reduce memory usage and increase throughput]_
+
+**Problem:** _[Installation of dependencies fails due to network issues]_
+
+**Solution:** _[Ensure your network connection is stable and consider using a proxy if necessary]_
+
+**Problem:** _[Python is not recognized as a command]_
+
+**Solution:** _[Make sure Python is installed and accessible in your system's `PATH`]_
+
+**Problem:** _[CUDA is not available after installation.]_
+
+**Solution:** _[Verify that your GPU supports CUDA and that the correct drivers are installed.]_
+
```

Comment:
A self review to myself:

Add Benchamarks and Performance of Model after and before GPU utilisation